content/blog/2022-03-21-conservative-gc/index.md

```diff
@@ -0,0 +1,256 @@
+against the object map, which is updated after every collection cycle.
+
+Similarly, **Conservative Sticky Immix (Sticky Immix<sub>Cons</sub>)** is the conservative
+counterpart of Sticky Immix. Maintaining the object map is trickier since only young objects
+are traced regularly.
+
+**Conservative Immix RC (RC Immix<sub>Cons</sub>)** performs tracing on young objects, and
+reference counting on mature objects. Object map is used here as well, and is updated incrementally
+as reference counting is performed.
+
+# Evaluation
+
+The authors conduct an extensive evaluation in two parts: (1)
+understanding the impact of conservatism by comparing between exact
+and conservative versions of collectors; (2) a performance evaluation
+comparing the conservative versions of RC, Immix, Sticky Immix, and RC
+Immix against existing state-of-the-art conservative garbage
+collectors.
+
+## Impact of Conservatism
+
+The authors evaluated the impacts of conservative-ness on collector
+mechanisms and design, specifically on the number of roots tracked,
+filtering, excess retention, and pinning.
+
+The takeaways are as follows:
+
+- *Ambiguous Pointers*: A challenge in conservative collection is how
+   often a non-pointer object may be interpreted as a pointer. The
+   authors found that conservative scanning results in _60%_
+   more "roots" than exact scanning.
+
+- *Excess Retention*: While excess retention is a very obvious side
+   effect of conservative collection, its precise impacts in practice
+   were not known. The authors measured excess retention by comparing
+   the sizes of transitive closures between the exact and conservative
+   versions, and found that excess retention was, on average, only _0.02%_,
+   with the maximum being only _6.1%_, as a fraction of heap size.
+
+- *Pointer Filtering*: The authors compare the performances
+  between their object map and the state-of-the-art BDW free-list
+  introspection, which are functionally equivalent ways of filtering
+  ambiguous roots. Object maps had a higher overhead in total,
+  mutator, and collection times primarily due to (1) setting bits at
+  allocation time, and (2) a space penalty that results from having to
+  store the map. The authors concluded that in the context of a
+  non-moving collector, BDW is clearly the better solution; however,
+  copying allows for a greater performance benefit.
+
+- *Pinning Granularity*: Conservative collectors need to pin the
+   "referents" of ambiguous pointers, but the effect of pinning
+   depends on the collector's pinning granularity. In MCCs, where
+   the "referent" and all other objects on the page that
+   it resides in are retained, _2.1%_ of the live heap was impacted.
+   In the 256 bytes line granularity (used by the Immix family of
+   collectors), where only the referent line is pinned, _0.2%_ of the
+   live heap was impacted. Therefore, the authors concluded that
+   pinning at the line granularity is significantly less impact-full.
```

Comment:
Pretty sure plain ol' "impactful" is a word.

Reply:
Fixed the typo!